### ISSUE TITLE
False positive unbound variables though all cases are checked with `isinstance`

### ISSUE THREAD

**Issue by Andrej730** [microsoft/pyright#8233]
False positive unbound variables, see example below:

```python
class A: ...
class B: ...

# works fine
def test_1(obj: A | B) -> None:
    if isinstance(obj, A):
        new_name = "test"
    elif isinstance(obj, B):
        new_name = "test"
    a = new_name


# breaks
def get_obj() -> A | B: ...
def test_2() -> None:
    obj = get_obj()
    if isinstance(obj, A):
        new_name = "test"
    elif isinstance(obj, B):
        new_name = "test"
    # "new_name" is possibly unbound
    a = new_name
```

**Comment by erictraut** [comment]
Pyright is working as designed here, so this isn't a bug.

See [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=narrowing-for-implied-else) for limitations on "implied else" narrowing.

**Comment by Andrej730** [comment]
Okay, I guess a function return type is too implicit for this :(
As a workaround `obj: A | B = get_obj()` seems to work fine.
